cpp/src/neighbors/detail/cagra/search_single_cta_kernel-inl.cuh

```diff
@@ -627,28 +669,27 @@ __device__ void search_core(
 
       // topk with bitonic sort
       _CLK_START();
-      if (std::is_same<SAMPLE_FILTER_T, cuvs::neighbors::filtering::none_sample_filter>::value ||
-          *filter_flag == 0) {
-        topk_by_bitonic_sort<MAX_ITOPK, MAX_CANDIDATES>(result_distances_buffer,
-                                                        result_indices_buffer,
-                                                        internal_topk,
-                                                        result_distances_buffer + internal_topk,
-                                                        result_indices_buffer + internal_topk,
-                                                        search_width * graph_degree,
-                                                        topk_ws,
-                                                        (iter == 0),
-                                                        multi_warps_1,
-                                                        multi_warps_2);
-        __syncthreads();
-      } else {
-        topk_by_bitonic_sort_1st<MAX_ITOPK + MAX_CANDIDATES>(
-          result_distances_buffer,
-          result_indices_buffer,
-          internal_topk + search_width * graph_degree,
-          internal_topk,
-          false);
+      if (!(std::is_same<SAMPLE_FILTER_T, cuvs::neighbors::filtering::none_sample_filter>::value ||
+            *filter_flag == 0)) {
+        // Move the filtered out index to the end of the itopk list
+        for (unsigned i = 0; i < search_width; i++) {
+          move_invalid_to_end_of_list(
+            result_indices_buffer, result_distances_buffer, internal_topk);
+        }
```

Comment:
Do I understand that right, that this algorithm moves one index at a time, and repeats this for each candidate in the list? That's `O(search_width*(parent_list_buffer + search_width))` complexity?

Maybe we'd better do a prefix scan (to get the indices of the valid items) followed by a shift of all elements (e.g. we can use `cub::WarpScan` or `cub::BlockScan` for that)? Or do you think the performance difference would be negligible?

Reply:
The reason I didn't use the scan method is that the maximum number of filtered-out elements here is `search_width`, which is typically 1. In our experience, increasing `search_width` does not help improve the search performance much, so it is likely to be a small number. So I use the specialized function for 1 here and run it multiple times if `search_width` is not 1.

Reply:
Makes sense, thanks for the explanation. For some reason I thought we could have more than `search_width` new  candidates each iteration.